h2o-algos/src/main/java/hex/tree/BuildHistogram.java

```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
       int sz = (int)(espc[i] - espc[i-1]);
       if(sz > largestChunkSz) largestChunkSz = sz;
     }
-    final int fLargestChunkSz = largestChunkSz;
     if(_weightIdx == -1){
       double [] ws = new double[largestChunkSz];
       Arrays.fill(ws,1);
       Arrays.fill(_ws,ws);
     }
-    final AtomicInteger cidx = new AtomicInteger(0);
-    // First do the phase 1 on all local data
-    new LocalMR(new MrFun(){
-      // more or less copied from ScoreBuildHistogram
-      private void map(int id, Chunk [] chks) {
-        final C4VolatileChunk nids = (C4VolatileChunk) chks[_nidIdx];
-        // Pass 1: Score a prior partially-built tree model, and make new Node
-        // assignments to every row.  This involves pulling out the current
-        // assigned DecidedNode, "scoring" the row against that Node's decision
-        // criteria, and assigning the row to a new child UndecidedNode (and
-        // giving it an improved prediction).
-        int [] nnids;
-        if( _leaf > 0)            // Prior pass exists?
-          nnids = score_decide(chks,nids.getValues());
-        else {                     // Just flag all the NA rows
-          nnids = new int[nids._len];
-          int [] is = nids.getValues();
-          for (int row = 0; row < nids._len; row++) {
-            if (isDecidedRow(is[row]))
-              nnids[row] = DECIDED_ROW;
-          }
-        }
-        // Pass 2: accumulate all rows, cols into histograms
-        // Sort the rows by NID, so we visit all the same NIDs in a row
-        // Find the count of unique NIDs in this chunk
-        int nh[] = (_nhs[id] = new int[_numLeafs + 1]);
-        for (int i : nnids)
-          if (i >= 0)
-            nh[i + 1]++;
-        // Rollup the histogram of rows-per-NID in this chunk
-        for (int i = 0; i <_numLeafs; i++) nh[i + 1] += nh[i];
-        // Splat the rows into NID-groups
-        int rows[] = (_rss[id] = new int[nnids.length]);
-        for (int row = 0; row < nnids.length; row++)
-          if (nnids[row] >= 0)
-            rows[nh[nnids[row]]++] = row;
+    new LocalMR(new PhaseOne(), new PhaseTwo(this, largestChunkSz)).fork();
+  }
+  
+  private class PhaseOne extends MrFun<PhaseOne> {
```

Comment:
It is a really nice change to have each pass as separate classes. However, I would prefer the classes have names that describe what they do. If not, add a brief docstring, please. 

I suggested something like ScoreAndSortPhase and BuildPhase or something like that.